Add tutorial code

sisyphusSmiling opened the pull request:
This tutorial is mean to cover Cadence attachments in the mental framework for which they're intended - as glue between resources to accomplish permissionless composability. Before continuing further, asking for review of the concept and overall walkthrough outlined in the README.

Largely inspired by the existing KittyVerse [Composable Resources ](https://developers.flow.com/cadence/tutorial/resources-compose) tutorial, this PR contains two primary NFT contracts - KittyVerse & KittyHats. As you might expect, KittyHats go on KittyVerse NFTs - cats on hats, if you will.

Considering the composable resources tutorial was meant to replace the KittyVerse tutorial, I thought refactoring the latter to feature attachments would be a good use of the idea so it remains in the dev onboarding funnel. However, I'm open to other ideas, especially gaming-focussed models such as [this one](https://github.com/sisyphusSmiling/rpg-attachments-example) or those building on existing NFT Collections.

FYI - once attachments are enabled on Playground, the plan is to move this tutorial to the Cadence tutorials found in onflow/docs.

sisyphusSmiling committed:
impl NFT in KittyHats

sisyphusSmiling committed:
add removal of KittyHats.NFT from HatAttachment

sisyphusSmiling committed:
fix contract bugs & update flow.json

sisyphusSmiling committed:
fix KittyHats contract bug

sisyphusSmiling committed:
add supporting txns & scripts

sisyphusSmiling committed:
add e2e tutorial scripts & transactions

sisyphusSmiling committed:
init README.md

sisyphusSmiling committed:
update KittyVerse comments

sisyphusSmiling committed:
update KittyHats comments

sisyphusSmiling committed:
update transaction comments

sisyphusSmiling committed:
add to README

satyamakgec commented on README.mdx:
```diff
@@ -0,0 +1,145 @@
+- What are they?
+    - In Cadence
+    - Simple mental model
+        - "Like glue between resources"
```
Don't you think we need more detailed documentation here? Maybe a link to a more detailed version of the documentation.

sisyphusSmiling replied:
Yes definitely, though it's just an outline for now. I wanted to get some eyes on where things are in code before diving too deep on documentation.

satyamakgec commented on README.mdx:
```diff
@@ -0,0 +1,145 @@
+        - As mentioned previously, attachments are like glue between resources
+    - As of now, attachments can be fully referenced anywhere the NFT can be referenced - this can complicate access 
+    control
+        - We made `removeHat()` `access(contract)` and callable in Collection only
```
Maybe a link for access modifiers documentation

satyamakgec commented on README.mdx:
```diff
@@ -0,0 +1,145 @@
+    - see that your hat is back in your Collection
+
+# Extra Credit
+
```
Don't you go to explain how the attachment works or whatnot from the code .. it looks like a walkthrough? Or am I missing something?

sisyphusSmiling replied:
I mentioned it above, but to clarify here - as it stands now it's just an outline of points to expand on atm. Was hoping to validate the code and the idea before investing more time in the documentation.

satyamakgec replied:
Okay cool make sense

satyamakgec commented on README.mdx:
```diff
@@ -0,0 +1,145 @@
+
+## Contract Overview
+
+- Contract interface implementation
```
I would suggest to give an overview about the design choices and give a disclaimer that it is up to the developer to use attachment for other purposes and they can have there own design choices but we recommend to follow the best practices of writing the smart contracts in cadence.

satyamakgec commented on README.mdx:
```diff
@@ -0,0 +1,145 @@
+
+- Contract interface implementation
+    - Events
+    - Nested resource definitions
```
Pointer - Maybe it will be good to add that with attachments it is something not possible to do so or if possible then what were the drawbacks of the same so a developer can easily quantify the benefits of the attachments.

satyamakgec commented on README.mdx:
```diff
@@ -0,0 +1,145 @@
+
+## Setup account & Mint KittyVerse NFT
+
+- send `tx_01_setup_kittyverse_account.cdc` signing with `0x02`
```
I am assuming that you are going to broaden the transactions as well. Like what they are doing and what are the important nuances in the transactions and scripts that a developer should keep remembering it. Maybe good to add some links of transaction-related documentation

sisyphusSmiling replied:
Yes definitely will expand on the details of the transaction similar to how other current tutorials walk through the details of each transaction.

joshuahannan commented on contracts/KittyHats.cdc:
```diff
@@ -8,38 +8,320 @@ import "KittyVerse"
+            // Withdraw the NFT we're going to put in the attachment
+            let hat  <- self.withdraw(withdrawID: hatID) as! @NFT
+
+            // Add the hat to the HatAttachment
+            withAttachment[HatAttachment]!.addHatNFT(<-hat)
```
not sure how you're indexing into `withattachmet` here since it is an @NFT, or what `HatAttachment` is

sisyphusSmiling replied:
Ah right, I'll need to expand on the syntax used for attachments here and in the walkthrough. You can access a resource's attachment's with `base[StaticType]` which returns `attachment?`. This definitely deserves more clarity, so thanks for pointing it out.

joshuahannan replied:
oh yeah I forgot about that. I guess I need to brush up on my attachment knowledge. 😆

joshuahannan reviewed (commented):
I think the code and the outline is a great start! Just remember that this might be some people's first exposure to the full NFT standard, so it might make sense to keep that in mind when explaining some concepts. You don't need to explain the whole standard, just be a little more thorough when talking about things that weren't already mentioned in a previous tutorial

sisyphusSmiling committed:
expand on README

sisyphusSmiling commented:
Thanks for the feedback all! Revisiting addressing some of the points made about the README.

@satyamakgec could you take another look, please? Hoping to get the outline and code approved before continuing on with a more thorough walkthrough in the README.

franklywatson commented on README.mdx:
```diff
@@ -0,0 +1,198 @@
+Composability is the ability for existing components in one construction to be reused and even extended for another 
+purpose. <br/>
+
+Think of composable architectures like [building with legos](https://twitter.com/cdixon/status/1450462593232089088?s=20) 
+- you build something you need as a block or set of blocks and someone else can come along and use those same blocks to 
+build something entirely new.
```
Maybe we shine a light on strengths Cadence specifically has for composability namely that contracts can compose other contracts and also load them dynamically on-demand. This makes working with contracts more similar to libraries than any other smart contract language and this education will help them see things more flexibly

sisyphusSmiling replied:
Good idea! Do we have docs that expand on this?

franklywatson replied:
Borrowing contracts shows it best: https://developers.flow.com/cadence/language/contracts#borrowing-a-deployed-contract

franklywatson commented on README.mdx:
```diff
@@ -0,0 +1,198 @@
+- you build something you need as a block or set of blocks and someone else can come along and use those same blocks to 
+build something entirely new.
+
+Composability on Flow, especially with attachments is permissionless because not only can you build with existing 
+components, but no one can stop you from doing so!
```
```suggestion
Composability on Flow, especially with attachments is permissionless because not only can you build with existing 
components, but no one can stop you from doing so! Hence the term permissionless!
```

franklywatson commented on README.mdx:
```diff
@@ -0,0 +1,198 @@
+
+</Callout>
+
+Permissionless composability is huge because, as builders, we can build on existing constructs for existing 
+communities and be sure that the original developer won't rug our access - AKA platform risk.
+
+For an example of the negative case, just look at how big tech companies locked down their API in the early days.
+
+Less duplicate code, easier to bootstrap attention and adoption, and users benefit from the utility you built into 
+something they already presumably own and love - win-win-win!
```
```suggestion
Permissionless composability is one of the unique characteristics of decentralized app development because, as builders, we can freely compose richer or derivative functionalities on top of any existing contracts from other applications or communities. Contracts that you deploy which interact with other contracts are not subject to internal changes or control by the contract you composed it upon. 

Less duplicate code, easier to bootstrap attention and adoption, and users benefit from the utility you built into 
something they already presumably own and love - win-win-win!
```

I struggle a little with the 'original developer won't rug our access' since for Web3 newcomers it might not be enough for them to make a connection. We may want to drop the point about platform risk - it's still the case that if you build to Flow you can't easily switch platforms whereas the same doesn't apply to EVM compatible equivalents. 

I'm also not sure there's enough strength in the reference to big tech APIs for the suggested wording. GCP copied AWS in terms of APIs to smooth migration, so in a sense the locked down aspects of the cloud provider APIs is their strength since they are so ubiquitously supported/similar

sisyphusSmiling replied:
> it's still the case that if you build to Flow you can't easily switch platforms whereas the same doesn't apply to EVM compatible equivalents.

Definitely a valid point. By "platform" here I mean something like Twitter's API when they proclaimed to be a protocol, not necessarily the the blockchain itself. The ground may at any point fall out from under you if the chain ceases to exist, but, taking the chain's existence as a given, I'd say the immutability guarantees of smart contracts provide a risk-reduced platform in comparison to privately owned and managed APIs. Case in point being the aforementioned Twitter API.

Regardless, your verbiage gets the point across more clearly for sure. Thanks for the rephrase!

franklywatson commented on README.mdx:
```diff
@@ -0,0 +1,198 @@
+    - Events
+    - Nested resource definitions
+- Attribute overview
+- Note on `unsafeRandom()`
```
Maybe leave this out for now since we're about to launch SafeRandom in this sprint cycle

franklywatson commented on README.mdx:
```diff
@@ -0,0 +1,198 @@
+    - Attachments are a special type that can be added to resources (doc link needed)
+    - You might be tempted to simply define all your functionality in the attachment, but you'll lose the existence and
+      uniqueness guarantees inherent to resources as well as make access control significantly harder for yourself
+        - As mentioned earlier, attachments are best thought of as glue between resources
```
Maybe we should be explicit and also say that they aren't Resources and shouldn't be thought of that way? Maybe also we should explain that it's utility is not for use when within ones own contracts, but when other contracts want to interact and contribute something new to an existing Resource

sisyphusSmiling replied:
> Maybe we should be explicit and also say that they aren't Resources and shouldn't be thought of that way?

Agreed, worth noting that fact

> Maybe also we should explain that it's utility is not for use when within ones own contracts, but when other contracts want to interact and contribute something new to an existing Resource

IDK if we need to be that prescriptive about their use. It's not clear to me that it's necessarily an anti-pattern to develop with attachments within contracts you own. Is there a reason you think it's bad practice?

satyamakgec replied:
It would be better not to have the attachment at the very first place. until I know this attachment would be useful to some x-audience. So it would be easy to manage and less error-prone. While we can add attachments at any time during upgrades, isn't it?

franklywatson replied:
> It's not clear to me that it's necessarily an anti-pattern to develop with attachments within contracts you own. Is there a reason you think it's bad practice

Not entirely sure myself, it's something that @joshuahannan spoke to more. He argued before that if it's your contract you don't need attachments since you can explicitly design the functionality into the main contract. Maybe we should make sure the documentation makes it clear that providing support for attachments in your Resources is a key way to enable composability from unknown future sources because it signals that the project is open to collaboration/contributions

sisyphusSmiling replied:
Definitely don't need attachments if it's your contract. That said, I don't know that we need to discourage that approach.

For instance, let's say I want to create a generic gaming NFT. I might create a base NFT and define my in-game resources as attachments. This means the access pattern across all of the in-game resources are consistent - for those that I defined and others that emerge in the future. I envision this being closest to token-bound accounts.

But I'll add in verbiage around the fact that attachments aren't necessary if you control the contracts.

I do think we can allow for some emergent patterns where an approach isn't apparently harmful, so hesitant to be too prescriptive here.

franklywatson commented on README.mdx:
````diff
@@ -0,0 +1,198 @@
+    // Otherwise, add the attachment
+    return <- attach HatAttachment() to <- toNFT
+    ```
+- Gotchas
+    - Accessing attachments of a given type requires a static type declaration
+    - Attempting to add an attachment to a resource when one already exists will result in a runtime error
+        - This is why we check if an attachment already exists before adding one in the example above
+        - You'll want to lean on simple helper methods to conditionally add attachments within your contracts
````
Maybe leave Gotcha's to a later section where we summarize what was done in the tutorial and add that color then?

sisyphusSmiling replied:
Good idea. I can move it to the latter part for reflection on aspects of the design choices

franklywatson commented on README.mdx:
```diff
@@ -0,0 +1,198 @@
+        - This is why we check if an attachment already exists before adding one in the example above
+        - You'll want to lean on simple helper methods to conditionally add attachments within your contracts
+- Great for
+    - Building on existing resources
```
```suggestion
    - Building on existing resources in other contracts 
```

sisyphusSmiling committed:
Update README.mdx

Co-authored-by: j pimmel <frankly.watson@gmail.com>

sisyphusSmiling committed:
Update README.mdx

Co-authored-by: j pimmel <frankly.watson@gmail.com>

sisyphusSmiling committed:
Add consideration to README

sisyphusSmiling committed:
add PR recommendations to README

sisyphusSmiling committed:
expand on composability in README

satyamakgec commented on README.mdx:
```diff
@@ -0,0 +1,206 @@
+this guidance to make sense.
+
+# KittyVerse + KittyHats
+
```
Do you think it would be good to add the reference of kittyVerse & kittyHats repo ? or a small context about them ?

satyamakgec commented on README.mdx:
```diff
@@ -0,0 +1,206 @@
+communities. Contracts that you deploy which interact with other contracts are not subject to internal changes or
+control by the contract you composed it upon. 
+
+Cadence is uniquely suited for composable architectures because Cadence contracts can compose other contracts. As a
```
I don't know what is the right way to put it but this line seems odd to me ` because Cadence contracts can compose other contracts`.  Maybe re-write like this `Cadence excels in facilitating composable architectures due to its ability to compose other contracts through Cadence contracts.`

satyamakgec commented on README.mdx:
````diff
@@ -0,0 +1,206 @@
+    ```cadence
+    access(self) fun addAttachment(toNFT: @KittyVerse.NFT): @KittyVerse.NFT {
+        // If attachment already exists, return
+        if toNFT[HatAttachment] != nil {
````
My assumption is that you are going to break it down this little bit. However, it is a readable code but newcomers may not understand the syntax `toNFT[HatAttachment]` as it only apply to attachments.

sisyphusSmiling replied:
Yes, I'll add more context and a general case before introducing contract associated verbiage.

satyamakgec commented on README.mdx:
```diff
@@ -0,0 +1,206 @@
+            - If KittyVerse already has an enthusiastic community behind it
+    - Let's say we're defining both, why use attachments?
+        - Building with attachments is a clear signal that you're designing for this permissionless composability
+            - You encourage others to define their own sorts of attachments for KittyVerse NFTs
```
is it possible to create attachments with changing the underlying contract ?

So let's say Alice creates a KittyVerse contract and it doesn't have any attachment functionality and bob owns one kittyVerse and wants to attach a kittyHat NFT to its KittyVerse NFT. Then bob can permisionlessely attach the kittyHat to its kittyVerse without the intervene of Alice changing the contract to add the support of attachments ?

sisyphusSmiling replied:
> is it possible to create attachments with changing the underlying contract ?
...
> bob can permisionlessely attach the kittyHat to its kittyVerse without the intervene of Alice changing the contract to add the support of attachments ?

Yes, that's correct. Bob would just create and deploy the KittyHats contract, and then he could add the KittyHats HatAttachment to his KittyVerse NFTs. No change needed by Alice to the KittyVerse contract

sisyphusSmiling committed:
expand on syntax in README
